runtime/secrets/reader.go

```diff
@@ -0,0 +1,192 @@
+		return nil, fmt.Errorf("failed to parse proxy address %q: %w", address, err)
+	}
+
+	username, hasUsername := secret.Data[UsernameKey]
+	password, hasPassword := secret.Data[PasswordKey]
+
+	if hasUsername && hasPassword {
+		proxyURL.User = url.UserPassword(string(username), string(password))
+	} else if hasUsername {
+		proxyURL.User = url.User(string(username))
+	}
+
+	return proxyURL, nil
+}
+
+// BasicAuthFromSecret retrieves basic authentication credentials from a Kubernetes secret.
+//
+// The function expects the secret to contain "username" and "password" fields.
+// Both fields are required and the function will return an error if either is missing.
+func BasicAuthFromSecret(ctx context.Context, c client.Client, name, namespace string) (string, string, error) {
+	secret, err := getSecret(ctx, c, name, namespace)
+	if err != nil {
+		return "", "", err
+	}
+
+	usernameData, exists := secret.Data[UsernameKey]
+	if !exists {
+		return "", "", fmt.Errorf("key %q not found in secret", UsernameKey)
+	}
+
+	passwordData, exists := secret.Data[PasswordKey]
+	if !exists {
+		return "", "", fmt.Errorf("key %q not found in secret", PasswordKey)
+	}
+
+	return string(usernameData), string(passwordData), nil
+}
+
+func PullSecretsFromServiceAccount(ctx context.Context, c client.Client, name, namespace string) ([]corev1.Secret, error) {
+	serviceAccount := &corev1.ServiceAccount{}
+	saKey := types.NamespacedName{Name: name, Namespace: namespace}
+	if err := c.Get(ctx, saKey, serviceAccount); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("serviceaccount %s/%s not found", namespace, name)
+		}
+		return nil, fmt.Errorf("failed to get serviceaccount %s/%s: %w", namespace, name, err)
+	}
+
+	if len(serviceAccount.ImagePullSecrets) == 0 {
+		return []corev1.Secret{}, nil
+	}
+
+	secrets := make([]corev1.Secret, 0, len(serviceAccount.ImagePullSecrets))
+	for _, imagePullSecret := range serviceAccount.ImagePullSecrets {
+		secret, err := getSecret(ctx, c, imagePullSecret.Name, namespace)
+		if err != nil {
+			return nil, err
+		}
+		secrets = append(secrets, *secret)
+	}
+
+	return secrets, nil
+}
+
+func getSecret(ctx context.Context, c client.Client, name, namespace string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	secretKey := types.NamespacedName{Name: name, Namespace: namespace}
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("secret %s/%s not found", namespace, name)
+		}
+		return nil, fmt.Errorf("failed to get secret %s/%s: %w", namespace, name, err)
+	}
+	return secret, nil
+}
+
+func getSecretData(secret *corev1.Secret, key, fallbackKey string, supportDeprecated bool) ([]byte, error) {
+	if data, exists := secret.Data[key]; exists {
+		return data, nil
+	}
+
+	if supportDeprecated {
+		if data, exists := secret.Data[fallbackKey]; exists {
+			return data, nil
+		}
+	}
+
+	return nil, fmt.Errorf("key %q not found in secret", key)
```

Comment:
nit: I'm not a big fan of `%q` :cry: I've debugged some JSON logs with `\"` in the middle of the string and it can be a bit annoying... I personally prefer `'%s'`